simulators/ethereum/pyspec/runner.go

```diff
@@ -0,0 +1,196 @@
+			return nil
+		}
+
+		// create testcase structure from fixtureTests
+		for name, fixture := range fixtureTests {
+			network := fixture.json.Network
+			// skip networks post merge
+			for _, skip := range []string{"Istanbul", "Berlin", "London"} {
+				if strings.Contains(network, skip) {
+					continue
+				}
+			}
+			// check network is valid 
+			if _, exist := envForks[network]; !exist {
+				return fmt.Errorf("network `%v` not defined in fork ruleset", network)
+			}
+
+			// extract genesis fields from fixture test
+			genesis := extractGenesis(fixture.json)
+
+			// extract payloads from each block
+			payloads := []*api.ExecutableData{}
+			for _, block := range fixture.json.Blocks {
+				gblock, _ := block.decodeBlock()
+				payload := api.BlockToExecutableData(gblock, common.Big0).ExecutionPayload
+				payloads = append(payloads, payload)
+			}
+
+			// extract post account information
+			postAlloc := fixture.json.Post
+
+			// define testcase struct adding the extracted fields
+			tc := testcase {
+				fixture: fixture, 
+				name: name, 
+				filepath: path,
+				genesis: genesis,  
+				payloads: payloads,
+				postAlloc: postAlloc,
+			}
+
+			// feed testcase to single worker within fixtureRunner()
+			fn(tc)
+		}
+		return nil
+
+	})
+}
+
+// run launches the client and runs the test case against it.
+func (tc *testcase) run(t *hivesim.T) {
+	start := time.Now()
+
+	t.Log("HIVE LOG --> Setting variables required for starting client.")
+	engineAPI := hive_rpc.HiveRPCEngineStarter{
+		ClientType: tc.clientType,
+		EnginePort: globals.EnginePortHTTP,
+		EthPort:    globals.EthPortHTTP,
+		JWTSecret:  globals.DefaultJwtTokenSecretBytes,
+	}
+	ctx := context.Background()
+	env := hivesim.Params{
+		"HIVE_FORK_DAO_VOTE": "1",
+		"HIVE_CHAIN_ID":      "1",
+		"HIVE_SKIP_POW":      "1",
+		"HIVE_NODETYPE": 	  "full",
+	}
+	tc.updateEnv(env)
+	t0 := time.Now()
+
+	// start client (also creates an engine RPC client internally)
+	t.Log("HIVE LOG --> Starting client with Engine API.")
+	engineClient, err := engineAPI.StartClient(t, ctx, tc.genesis, env, nil)
+	if err != nil {
+		t.Fatalf("HIVE FATAL --> Can't start client with Engine API: %v", err)
+	}
+	t1 := time.Now()
+
+	// send payloads
+	for blockNumber, payload :=  range tc.payloads {
+		_, err := engineClient.NewPayloadV2(context.Background(), payload)
+		if err != nil {
+			t.Fatalf("HIVE FATAL ---> Unable to send payload[%v] in test %s: %v ", blockNumber, tc.name, err)
+		}
+	}
+	t2 := time.Now()
+
+	// update head of "beacon" chain with the latest valid response
+	latestBlockHash := *engineClient.LatestNewPayloadResponse().LatestValidHash
+	fcState := &api.ForkchoiceStateV1{HeadBlockHash: latestBlockHash}
+	engineClient.ForkchoiceUpdatedV2(ctx, fcState, nil)
+	t3 := time.Now()
+
+	// check nonce, balance & storage of accounts in final block against fixture values
+	latestBlockNumber := new(big.Int).SetUint64(engineClient.LatestNewPayloadSent().Number)
```

Comment:
It can be the case that the latest payload was `INVALID`, because the test is verifying that an invalid block is rejected.
In that case, `engineClient.LatestNewPayloadSent().Number` will be an unknown block to the client.
It should be ok to simply use `nil` instead of `latestBlockNumber`, to always check the head of the chain after the blocks were executed.